Basic/ClientDriven/Assets/Scripts/NetworkObjectSpawner.cs

```diff
@@ -0,0 +1,38 @@
+using System;
+using Unity.Netcode;
+using UnityEngine;
+using UnityEngine.Networking;
+using Random = System.Random;
+
+namespace Unity.Multiplayer.Samples.ClientDriven
+{
+    public class NetworkObjectSpawner : MonoBehaviour
+    {
+        [SerializeField] NetworkObject prefabReference;
+
+        public void Start()
+        {
+            if (NetworkManager.Singleton != null)
```

Comment:
Are you 100% sure the NetworkKamanger exists when you check this?
Since this is a 1-time check, if the NetworkManager is not instanciated, this class will do nothing.

Ideally, you should wait until the NetworkManager is instanciated, using a coroutine.

Reply:
I used a Debug.Assert instead now

Reply:
Debug.Assert is basically a one-time if, it doesn't solve the issue.

> It never threw an error, so I think it should be fine?

If the application flow is exactly the same every time, it's fine. But for exampel you could have race conditions if you reconnect vs when you start the first time.

This is what you need to be sure this works 100% of times: https://docs.unity3d.com/ScriptReference/WaitUntil.html

Reply:
I will try to insert a coroutine then

Reply:
Are we sure that a coroutine is really necessary here? Considering that this will be attached to (potentially) many objects, having that many coroutines will be costly on initial startup for a scene.

IMO it's okay that this kind of script is opinionated and expects a NetworkManager to exist in the same scene as this GameObject. That can be communicated in this class' summary.

Reply:
@RikuTheFuffs

Reply:
In which way would insert a coroutine, would it lead to waiting until NetworkManager.Singleton is not null?

Reply:
Is this the correct way to do it?: 

```
void Start()
            {

                if (NetworkManager.Singleton == null)
                {
                    StartCoroutine(NetworkManagerCoroutine());
                }

                if (NetworkManager.Singleton != null)
                {
                    NetworkManager.Singleton.OnServerStarted += OnServerStartedIngredientSpawn;
                }
            }

            IEnumerator NetworkManagerCoroutine()
            {
                Debug.Log("NetworkManager not here...");
                yield return new WaitUntil(() => NetworkManager.Singleton != null);
                Debug.Log("NetworkManager is here!");
            }
```

Reply:
Basically, I feel it’s okay to have either a direct reference to NetworkManager via the scene, or reference the NetworkManager’s Singleton.

I don’t see the need for a coroutine, since the following is true:

> If the application flow is exactly the same every time, it’s fine.

The script can be opinionated here in assuming that the scene flow won’t be different as there’s only one scene. Our dependency injection use-case inside of Boss Room is an example of handling in-scene dependencies.

And:

> But for example you could have race conditions if you reconnect vs when you start the first time.

NetworkManager.Singleton is assigned on OnEnable, and this component’s Start method is going to execute after that’s been assigned, and so I don’t see what race condition could come up here. Is there another I'm missing?

Lastly, adding a coroutine to objects that may have many, many instances in the scene is imo not a scalable recommendation. Coroutine usage should be done sparingly.

Let me know your thoughts.

Reply:
> I feel it’s okay to have either a direct reference to NetworkManager via the scene

That only works if the NetworkManager is in the same scene as these objects, and requires a lot of manual work if there are many instances of this class.

> or reference the NetworkManager’s Singleton.

Yes, but it needs to be initialized first.

> I don’t see the need for a coroutine, since the following is true:
> 
> > If the application flow is exactly the same every time, it’s fine.
> 
> The script can be opinionated here in assuming that the scene flow won’t be different as there’s only one scene. 

No problem then.

> NetworkManager.Singleton is assigned on OnEnable, and this component’s Start method is going to execute after that’s been assigned, and so I don’t see what race condition could come up here. Is there another I'm missing?

For now, in-editor. Is the same in a build?
Is it going to be the same in the next NGO update(s)?

There were many episodes in Unity's history where the order of events in-editor and in-build differed, especially with netcode frameworks.

> Lastly, adding a coroutine to objects that may have many, many instances in the scene is imo not a scalable recommendation. Coroutine usage should be done sparingly.

1. Have you profiled it? If not, better not to do assumptions.
2. A performance hiccup of a few frames in a non-critical gameplay situation (I.E: scene initialization) is totally fine if it is necessary to ensure the initialization flow works correctly. There's no point in having a script that works half of the time, especially if we claim these samples are "production-ready" and "copy-pastable".

Reply:
A couple of sanity checks on my end:
* Order of execution is the same on builds or editor, even when rehosting or rejoining.
* Profiled 100 spawners. While you don't notice a large effect on CPU usage (< 1ms with 100 objects), this is going to significantly add to an application's start up / scene load time, where you'd see 24 KB of memory allocated (& 1.37 ms to generate these coroutines, which can be crucial for scene loads and/or lower-end devices.

![Screenshot 2024-04-07 194450](https://github.com/Unity-Technologies/com.unity.multiplayer.samples.bitesize/assets/75813458/5483f133-7451-4fa0-8933-28af8d014fd5)

So, since execution order is consistent, this coroutine won't ever execute. Knowing so, my recommendation would be to use NetworkManager.Singleton which we know is assigned in this single-scene sample.

My comments weren't targeted at Coroutine usage, but moreso on us writing code that fits the sample's design. We can't be accounting for all the ways a user can re-arrange this sample.

Reply:
> While you don't notice a large effect on CPU usage (< 1ms with 100 objects), this is going to significantly add to an application's start up / scene load time, where you'd see 24 KB of memory allocated (& 1.37 ms to generate these coroutines, which can be crucial for scene loads and/or lower-end devices.)

Looking at your profiling capture I'd be more worried about the other 0.79 MB of GC allocated (especially the portion that comes from UI, since that will likely be there every frame).

> So, since execution order is consistent, this coroutine won't ever execute.

Great, so let's consider this matter resolved.